src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs

```diff
@@ -0,0 +1,932 @@
+            }
+            else
+            {
+                Accumulate(acc, input, secret.Slice((int)stripesSoFar * SecretConsumeRateBytes), (int)stripes);
+                stripesSoFar += stripes;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ulong MergeAccumulators(Span<ulong> acc, ReadOnlySpan<byte> secret, ulong start)
+        {
+            Debug.Assert(acc.Length == 8);
+            Debug.Assert(secret.Length >= 64);
+
+            ulong result64 = start;
+
+            result64 += Multiply64To128ThenFold(
+                acc[0] ^ BinaryPrimitives.ReadUInt64LittleEndian(secret.Slice(0)),
+                acc[1] ^ BinaryPrimitives.ReadUInt64LittleEndian(secret.Slice(8)));
+
+            result64 += Multiply64To128ThenFold(
+                acc[2] ^ BinaryPrimitives.ReadUInt64LittleEndian(secret.Slice(16)),
+                acc[3] ^ BinaryPrimitives.ReadUInt64LittleEndian(secret.Slice(24)));
+
+            result64 += Multiply64To128ThenFold(
+                acc[4] ^ BinaryPrimitives.ReadUInt64LittleEndian(secret.Slice(32)),
+                acc[5] ^ BinaryPrimitives.ReadUInt64LittleEndian(secret.Slice(40)));
+
+            result64 += Multiply64To128ThenFold(
+                acc[6] ^ BinaryPrimitives.ReadUInt64LittleEndian(secret.Slice(48)),
+                acc[7] ^ BinaryPrimitives.ReadUInt64LittleEndian(secret.Slice(56)));
+
+            return Avalanche(result64);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ulong Mix16Bytes(ReadOnlySpan<byte> source, ulong secretLow, ulong secretHigh, ulong seed)
+        {
+            Debug.Assert(source.Length >= 16);
+
+            ulong inputLow = BinaryPrimitives.ReadUInt64LittleEndian(source);
+            ulong inputHigh = BinaryPrimitives.ReadUInt64LittleEndian(source.Slice(8));
+
+            return Multiply64To128ThenFold(
+                inputLow ^ (secretLow + seed),
+                inputHigh ^ (secretHigh - seed));
+        }
+
+        /// <summary>Calculates a 32-bit to 64-bit long multiply.</summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ulong Multiply32To64(ulong v1, ulong v2) => (uint)v1 * (ulong)(uint)v2;
+
+        /// <summary>"This is a fast avalanche stage, suitable when input bits are already partially mixed."</summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ulong Avalanche(ulong hash)
+        {
+            hash = XorShift(hash, 37);
+            hash *= 0x165667919E3779F9;
+            hash = XorShift(hash, 32);
+            return hash;
+        }
+
+        /// <summary>Calculates a 64-bit to 128-bit multiply, then XOR folds it.</summary>
+        private static ulong Multiply64To128ThenFold(ulong lhs, ulong rhs)
+        {
+#if NET5_0_OR_GREATER
+            ulong upper = Math.BigMul(lhs, rhs, out ulong lower);
```

Comment:
Ouch, codegen for Math.BigMul doesn't look nice:
```asm
; Method System.IO.Hashing.XxHash3:Multiply64To128ThenFold(ulong,ulong):ulong
G_M64846_IG01:              ;; offset=0000H
       50                   push     rax
       4889542418           mov      qword ptr [rsp+18H], rdx
						;; size=6 bbWeight=1    PerfScore 2.00

G_M64846_IG02:              ;; offset=0006H
       488D0424             lea      rax, bword ptr [rsp]
       4C8B442418           mov      r8, qword ptr [rsp+18H]
       488BD1               mov      rdx, rcx
       C4C2B3F6D0           mulx     rdx, r9, r8
       4C8908               mov      qword ptr [rax], r9
       488BC2               mov      rax, rdx
       488B1424             mov      rdx, qword ptr [rsp]
       4833C2               xor      rax, rdx
						;; size=30 bbWeight=1    PerfScore 7.25

G_M64846_IG03:              ;; offset=0024H
       4883C408             add      rsp, 8
       C3                   ret      
						;; size=5 bbWeight=1    PerfScore 1.25
; Total bytes of code: 41
```
must be multi-reg return issue I assume?

Reply:
Much better on arm64:
```asm
; Method System.IO.Hashing.XxHash3:Multiply64To128ThenFold(ulong,ulong):ulong
G_M64846_IG01:              ;; offset=0000H
        A9BF7BFD          stp     fp, lr, [sp, #-0x10]!
        910003FD          mov     fp, sp
						;; size=8 bbWeight=1    PerfScore 1.50

G_M64846_IG02:              ;; offset=0008H
        9B017C02          mul     x2, x0, x1
        9BC17C00          umulh   x0, x0, x1
        CA000040          eor     x0, x2, x0
						;; size=12 bbWeight=1    PerfScore 7.50

G_M64846_IG03:              ;; offset=0014H
        A8C17BFD          ldp     fp, lr, [sp], #0x10
        D65F03C0          ret     lr
						;; size=8 bbWeight=1    PerfScore 2.00
```

Reply:
It turns out there is already a CQ issue for this: https://github.com/dotnet/runtime/issues/75594 just wanted to let you know that it might eat some of the perf from your PR on x64

Reply:
Thanks.  It's not worse than the fallback, though, right? i.e. there's nothing I should do differently now for this.

Reply:
@stephentoub I tried locally and yeah, there is nothing we currently can do about it unfortunately, this has to be fixed in JIT, I guess it can be an additional motivation to do it sooner. Should still be faster than the fallback I think

Reply:
I just looked at the issue.  The cited poor code gen there is for BigMul with longs, but this is using ulongs.  Is the code gen for ulongs deficient as well?  The issue suggests it's ok.